From 3f2d453bacb2b38b4945b0bb13f78b2528e8912e Mon Sep 17 00:00:00 2001
From: Sam Day <me@samcday.com>
Date: Sun, 27 Sep 2020 18:33:48 +0200
Subject: [PATCH] [qubesdb-daemon] support socket fd passing from systemd

---
 daemon/db-daemon.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/daemon/db-daemon.c b/daemon/db-daemon.c
index d2bae6f..05a60a0 100644
--- a/daemon/db-daemon.c
+++ b/daemon/db-daemon.c
@@ -482,6 +482,13 @@ int mainloop(struct db_daemon_data *d) {
 
 #define MAX_FILE_PATH 256
 int init_server_socket(struct db_daemon_data *d) {
+#if HAVE_SYSTEMD
+    if (sd_listen_fds(0) == 1) {
+        d->socket_fd = SD_LISTEN_FDS_START + 0;
+        return 1;
+    }
+#endif
+
     char socket_address[MAX_FILE_PATH];
     struct sockaddr_un sockname;
     int s;
-- 
2.28.0

