From 5e9a0c944f85f530b3ba30662557934ecf66f1e0 Mon Sep 17 00:00:00 2001
From: Sam Day <me@samcday.com>
Date: Tue, 13 Oct 2020 13:30:25 +0200
Subject: [PATCH] [qubesd] add service depends on libvirt virtxend.socket

---
 linux/systemd/qubesd.service | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/linux/systemd/qubesd.service b/linux/systemd/qubesd.service
index 3726f141..4a0fc6e2 100644
--- a/linux/systemd/qubesd.service
+++ b/linux/systemd/qubesd.service
@@ -1,6 +1,8 @@
 [Unit]
 Description=Qubes OS daemon
 Before=systemd-user-sessions.service
+Requires=virtxend.socket
+After=virtxend.socket
 
 [Service]
 Environment=SYSTEMD_DAEMONS=1
-- 
2.28.0

